README.md
# Phaser 3 TypeScript Project Template

This quick-start project template combines Phaser 3 with [TypeScript](https://www.typescriptlang.org/) and uses [Rollup](https://rollupjs.org) for bundling.

## Requirements

[Node.js](https://nodejs.org) is required to install dependencies and run scripts via `yarn`.

## Available Commands

| Command | Description |
|---------|-------------|
| `yarn install` | Install project dependencies |
| `yarn start` | Build project and open web server running project, watching for changes |
| `yarn build` | Builds code bundle with production settings (minification, no source maps, etc..) |

## Writing Code

After cloning the repo, run `yarn install` from your project directory. Then, you can start the local development
server by running `yarn run watch`. The first time you run this you should see the following demo run:

![Screenshot](screenshot.png "Phaser 3 Example")

After starting the development server with `yarn run watch`, you can edit any files in the `src` folder
and Rollup will automatically recompile and reload your server (available at `http://localhost:10001`
by default).

## Configuring Rollup

* Edit the file `rollup.config.dev.js` to edit the development build.
* Edit the file `rollup.config.dist.js` to edit the distribution build.

You will find lots of comments inside the rollup config files to help you do this.

Note that due to the build process involved, it can take around 20 seconds to build the initial bundle. Times will vary based on CPU and local drive speeds. The development config does not minify the code in order to save build time, but it does generate source maps. If you do not require these, disable them in the config to speed it up further.



CHANGELOG.md
# Phaser3 typescript boilerplate

## v1.0.4
- Add linting.
- Add robots.txt

## v1.0.3
- Use rollup-plugin-terser instead of rollup-plugin-uglify.
- Use yarn instead of npm.
- Add types in game.ts
- Add favicon.ico
- Remove game built files from /dist.


src/game.ts
import 'phaser';

export default class Demo extends Phaser.Scene {
  constructor() {
    super('Demo');
  }

  preload(): void {
    this.load.image('logo', 'assets/phaser3-logo.png');
    this.load.image('libs', 'assets/libs.png');
    this.load.glsl('bundle', 'assets/plasma-bundle.glsl.js');
    this.load.glsl('stars', 'assets/starfields.glsl.js');
  }

  create(): void {
    this.add.shader('RGB Shift Field', 0, 0, 800, 600).setOrigin(0);
    this.add.shader('Plasma', 0, 412, 800, 172).setOrigin(0);
    this.add.image(400, 300, 'libs');

    const logo: Phaser.GameObjects.Image = this.add.image(400, 70, 'logo');

    this.tweens.add({
      targets: logo,
      y: 350,
      duration: 1500,
      ease: 'Sine.inOut',
      yoyo: true,
      repeat: -1,
    });
  }
}

const config: Phaser.Types.Core.GameConfig = {
  type: Phaser.AUTO,
  backgroundColor: '#125555',
  width: 800,
  height: 600,
  scene: Demo,
};

const game: Phaser.Game = new Phaser.Game(config);


rollup.config.dist.js
import commonjs from '@rollup/plugin-commonjs';
import resolve from '@rollup/plugin-node-resolve';
import replace from '@rollup/plugin-replace';
import { terser } from 'rollup-plugin-terser';
import typescript from 'rollup-plugin-typescript2';

export default {

  //  Our games entry point (edit as required)
  input: [
    './src/game.ts'
  ],

  //  Where the build file is to be generated.
  //  Most games being built for distribution can use iife as the module type.
  //  You can also use 'umd' if you need to ingest your game into another system.
  //  The 'intro' property can be removed if using Phaser 3.21 or above. Keep it for earlier versions.
  output: {
    file: './dist/game.js',
    name: 'MyGame',
    format: 'iife',
    sourcemap: false,
    intro: 'var global = window;'
  },

  plugins: [

    //  Toggle the booleans here to enable / disable Phaser 3 features:
    replace({
      'typeof CANVAS_RENDERER': JSON.stringify(true),
      'typeof WEBGL_RENDERER': JSON.stringify(true),
      'typeof EXPERIMENTAL': JSON.stringify(true),
      'typeof PLUGIN_CAMERA3D': JSON.stringify(false),
      'typeof PLUGIN_FBINSTANT': JSON.stringify(false),
      'typeof FEATURE_SOUND': JSON.stringify(true)
    }),

    //  Parse our .ts source files
    resolve({
      extensions: [ '.ts', '.tsx' ]
    }),

    //  We need to convert the Phaser 3 CJS modules into a format Rollup can use:
    commonjs({
      include: [
        'node_modules/eventemitter3/**',
        'node_modules/phaser/**'
      ],
      exclude: [ 
        'node_modules/phaser/src/polyfills/requestAnimationFrame.js'
      ],
      sourceMap: false,
      ignoreGlobal: true
    }),

    //  See https://www.npmjs.com/package/rollup-plugin-typescript2 for config options
    typescript(),

    //  See https://www.npmjs.com/package/rollup-plugin-terser for config options
    terser({
      mangle: false
    })

  ]
};

rollup.config.dev.js
import commonjs from '@rollup/plugin-commonjs';
import resolve from '@rollup/plugin-node-resolve';
import replace from '@rollup/plugin-replace';
import serve from 'rollup-plugin-serve';
import typescript from 'rollup-plugin-typescript2';

export default {

  //  Our games entry point (edit as required)
  input: [
    './src/game.ts'
  ],

  //  Where the build file is to be generated.
  //  Most games being built for distribution can use iife as the module type.
  //  You can also use 'umd' if you need to ingest your game into another system.
  //  The 'intro' property can be removed if using Phaser 3.21 or above. Keep it for earlier versions.
  output: {
    file: './dist/game.js',
    name: 'MyGame',
    format: 'iife',
    sourcemap: true,
    intro: 'var global = window;'
  },

  plugins: [

    //  Toggle the booleans here to enable / disable Phaser 3 features:
    replace({
      'typeof CANVAS_RENDERER': JSON.stringify(true),
      'typeof WEBGL_RENDERER': JSON.stringify(true),
      'typeof EXPERIMENTAL': JSON.stringify(true),
      'typeof PLUGIN_CAMERA3D': JSON.stringify(false),
      'typeof PLUGIN_FBINSTANT': JSON.stringify(false),
      'typeof FEATURE_SOUND': JSON.stringify(true)
    }),

    //  Parse our .ts source files
    resolve({
      extensions: [ '.ts', '.tsx' ]
    }),

    //  We need to convert the Phaser 3 CJS modules into a format Rollup can use:
    commonjs({
      include: [
        'node_modules/eventemitter3/**',
        'node_modules/phaser/**'
      ],
      exclude: [ 
        'node_modules/phaser/src/polyfills/requestAnimationFrame.js'
      ],
      sourceMap: true,
      ignoreGlobal: true
    }),

    //  See https://www.npmjs.com/package/rollup-plugin-typescript2 for config options
    typescript(),

    //  See https://www.npmjs.com/package/rollup-plugin-serve for config options
    serve({
      open: true,
      contentBase: 'dist',
      host: 'localhost',
      port: 10001,
      headers: {
        'Access-Control-Allow-Origin': '*'
      }
    })

  ]
};